File: src/main/java/io/github/chindeaone/collectiontracker/gui/overlays/DeployableOverlay.java
package io.github.chindeaone.collectiontracker.gui.overlays;

import io.github.chindeaone.collectiontracker.config.ConfigAccess;
import io.github.chindeaone.collectiontracker.config.core.Position;
import io.github.chindeaone.collectiontracker.utils.HypixelUtils;
import io.github.chindeaone.collectiontracker.utils.parser.DeployableParser;
import io.github.chindeaone.collectiontracker.utils.rendering.RenderUtils;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.Font;
import net.minecraft.client.gui.GuiGraphics;

import java.util.ArrayList;
import java.util.List;

public class DeployableOverlay implements AbstractOverlay {

    private final Position position = ConfigAccess.getDeployablePosition();
    private final List<String> deployableLines = new ArrayList<>();
    private boolean renderingAllowed = true;

    @Override
    public String overlayLabel() {
        return "Mining Deployable Overlay";
    }

    @Override
    public Position position() {
        return this.position;
    }

    @Override
    public boolean isEnabled() {
        return ConfigAccess.isDeployableEnabled() && HypixelUtils.isOnSkyblock();
    }

    @Override
    public boolean isRenderingAllowed() {
        return renderingAllowed;
    }

    @Override
    public void setRenderingAllowed(boolean allowed) {
        this.renderingAllowed = allowed;
    }

    @Override
    public void render(GuiGraphics context) {
        if (!isEnabled()) return;
        List<String> lines = getDeployableLines();

        if (lines.isEmpty()) return;

        RenderUtils.drawOverlayFrame(context, position, () ->
                RenderUtils.renderStrings(context, lines)
        );
    }

    @Override
    public void updateDimensions() {
        if (!isEnabled()) return;
        List<String> lines = getDeployableLines();
        if (lines.isEmpty()) return;

        Font fr = Minecraft.getInstance().font;
        int maxW = 0;
        for (String l : lines) maxW = Math.max(maxW, fr.width(l));
        int h = fr.lineHeight * lines.size();

        position.setDimensions(maxW, h);
    }

    private List<String> getDeployableLines() {
        deployableLines.clear();

        String buff = DeployableParser.getBuff();
        String expireTime = DeployableParser.getRemainingTime();

        if (buff.isEmpty() || expireTime.isEmpty() || !DeployableParser.isNear()) {
            return deployableLines;
        }

        int timeLeft = 0;
        try {
            timeLeft = Integer.parseInt(expireTime.replace("s", ""));
        } catch (NumberFormatException ignored) {}

        String buffColor = DeployableParser.getBuffColor();
        if (timeLeft <= 5) {
            deployableLines.add(buffColor + buff + " §cSoon!");
        } else {
            deployableLines.add(buffColor + buff + " §e" + timeLeft + "s");
        }
        return deployableLines;
    }
}

File: src/main/java/io/github/chindeaone/collectiontracker/gui/overlays/DeployableOverlay.kt
package io.github.chindeaone.collectiontracker.gui.overlays

import io.github.chindeaone.collectiontracker.ModLoader
import io.github.chindeaone.collectiontracker.config.ConfigAccess.getDeployablePosition
import io.github.chindeaone.collectiontracker.config.ConfigAccess.isDeployableEnabled
import io.github.chindeaone.collectiontracker.config.core.Position
import io.github.chindeaone.collectiontracker.utils.parser.DeployableParser.buff
import io.github.chindeaone.collectiontracker.utils.parser.DeployableParser.buffColor
import io.github.chindeaone.collectiontracker.utils.parser.DeployableParser.isNear
import io.github.chindeaone.collectiontracker.utils.parser.DeployableParser.remainingTime

class DeployableOverlay : AbstractOverlay() {
    private var cachedLines: List<String> = emptyList()

    override val overlayLabel: String = "Lantern Deployable"

    override val position: Position get() = getDeployablePosition()

    override val isEnabled: Boolean get() = isDeployableEnabled()

    override fun updateDimensions() {
        if (!isEnabled) return
        updateLinesIfNeeded()

        super.updateDimensions()
    }

    override val lines: List<String>
        get() {
            updateLinesIfNeeded()
            return cachedLines
        }

    private fun updateLinesIfNeeded() {
        if (!isEnabled) {
            cachedLines = emptyList()
            return
        }

        if (ModLoader.clientTicks % 5L != 0L) return

        val currentBuff = buff
        val expireTime = remainingTime
        val currentNear = isNear
        val currentBuffColor = buffColor

        if (currentBuff.isEmpty() || expireTime.isEmpty() || !currentNear) {
            cachedLines = emptyList()
            return
        }

        var timeLeft: Int
        try {
            timeLeft = expireTime.replace("s", "").toInt()
        } catch (_: NumberFormatException) {
            cachedLines = emptyList()
            return
        }

        val newLines = if (timeLeft <= 5) {
            listOf("$currentBuffColor$currentBuff §cSoon!")
        } else {
            listOf("$currentBuffColor$currentBuff §e${timeLeft}s")
        }

        cachedLines = newLines
    }
}
